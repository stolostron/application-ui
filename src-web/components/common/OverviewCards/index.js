/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import React from 'react'
import { connect } from 'react-redux'
import { withLocale } from '../../../providers/LocaleProvider'
import { ArrowRightIcon } from '@patternfly/react-icons'
import {
  Accordion,
  AccordionItem,
  AccordionContent,
  AccordionToggle,
  Alert,
  Button,
  Skeleton
} from '@patternfly/react-core'
import resources from '../../../../lib/shared/resources'
import msgs from '../../../../nls/platform.properties'
import config from '../../../../lib/shared/config'
import {
  getSearchLinkForOneApplication,
  getAppOverviewCardsData
} from '../ResourceOverview/utils'
import ChannelLabels from '../ChannelLabels'
import TimeWindowLabels from '../TimeWindowLabels'
import { getClusterCount } from '../../../../lib/client/resource-helper'
import { REQUEST_STATUS } from '../../../actions'
import _ from 'lodash'

/* eslint-disable react/prop-types */

resources(() => {
  require('./style.scss')
})

const mapStateToProps = state => {
  const { HCMApplicationList, topology } = state
  return {
    HCMApplicationList,
    topology
  }
}

class OverviewCards extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      showSubCards: false,
      showDetailsCard: true
    }
  }

  UNSAFE_componentWillMount() {
    // update cards every 1s to pick up side-effect in
    // redux state (calculation of node statuses) created by
    // topology code
    const intervalId = setInterval(this.reload.bind(this), 1000)
    this.setState({ intervalId: intervalId })
  }

  reload() {
    this.setState(prevState => ({ pollToggle: !prevState.pollToggle }))
  }

  componentWillUnmount() {
    clearInterval(this.state.intervalId)
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (
      nextState.pollToggle === this.state.pollToggle ||
      _.get(nextProps, 'topology.status', '') === REQUEST_STATUS.IN_PROGRESS ||
      _.get(nextProps, 'HCMApplicationList.status', '') ===
        REQUEST_STATUS.IN_PROGRESS
    ) {
      return false
    }
    return true
  }

  render() {
    const {
      HCMApplicationList,
      topology,
      selectedAppName,
      selectedAppNS,
      locale
    } = this.props
    const { showSubCards, showDetailsCard } = this.state

    if (HCMApplicationList.status === REQUEST_STATUS.ERROR) {
      const errMessage = _.get(
        HCMApplicationList,
        'err.message',
        msgs.get('resource.error')
      )
      return (
        <Alert
          variant="danger"
          title={errMessage}
          className="overview-notification"
        />
      )
    }
    if (HCMApplicationList.status === REQUEST_STATUS.NOT_FOUND) {
      const infoMessage = _.get(
        HCMApplicationList,
        'err.err',
        msgs.get('load.app.info.notfound', [selectedAppName])
      )
      return (
        <Alert
          variant="info"
          title={infoMessage}
          className="overview-notification"
        />
      )
    }

    const toggleAccordion = toggleStatus => {
      if (toggleStatus) {
        this.setState({ showDetailsCard: false })
      } else {
        this.setState({ showDetailsCard: true })
      }
      this.forceUpdate()
    }

    let getUrl = window.location.href
    getUrl = getUrl.substring(0, getUrl.indexOf('/multicloud/applications/'))

    const targetLink = getSearchLinkForOneApplication({
      name: encodeURIComponent(selectedAppName),
      namespace: encodeURIComponent(selectedAppNS)
    })

    const appOverviewCardsData = getAppOverviewCardsData(
      HCMApplicationList,
      topology,
      selectedAppName,
      selectedAppNS,
      targetLink,
      locale
    )

    const clusterCount = getClusterCount({
      locale,
      remoteCount: appOverviewCardsData.remoteClusterCount,
      localPlacement: appOverviewCardsData.localClusterDeploy,
      name: selectedAppName,
      namespace: selectedAppNS,
      kind: 'application',
      apigroup: 'app.k8s.io'
    })

    const disableBtn =
      appOverviewCardsData.subsList &&
      appOverviewCardsData.subsList !== -1 &&
      appOverviewCardsData.subsList.length > 0
        ? false
        : true

    return (
      <div className="overview-cards-container">
        <Accordion>
          <AccordionItem>
            <AccordionToggle
              onClick={() => {
                toggleAccordion(showDetailsCard)
              }}
              isExpanded={showDetailsCard}
              className="overview-cards-details-section"
              id="details-card"
            >
              {msgs.get('dashboard.card.overview.cards.title', locale)}
            </AccordionToggle>

            <AccordionContent isHidden={!showDetailsCard}>
              <div className="details-col  add-right-border" id="left-col">
                <div className="details-item">
                  <div className="details-item-title left-item">
                    {msgs.get('dashboard.card.overview.cards.name', locale)}
                  </div>
                  <div className="details-item-content">
                    {appOverviewCardsData.appName}
                  </div>
                </div>

                <div className="details-item">
                  <div className="details-item-title left-item">
                    {msgs.get(
                      'dashboard.card.overview.cards.namespace',
                      locale
                    )}
                  </div>
                  <div className="details-item-content">
                    {appOverviewCardsData.appNamespace}
                  </div>
                </div>

                <div className="details-item">
                  <div className="details-item-title left-item">
                    {msgs.get('dashboard.card.overview.cards.created', locale)}
                  </div>
                  <div className="details-item-content">
                    {this.renderData(
                      appOverviewCardsData.creationTimestamp,
                      appOverviewCardsData.creationTimestamp,
                      '30%'
                    )}
                  </div>
                </div>
              </div>

              <div className="details-col" id="righ-col">
                <div className="details-item">
                  <div className="details-item-title right-item">
                    {msgs.get('dashboard.card.overview.cards.clusters', locale)}
                  </div>
                  <div className="details-item-content">
                    {this.renderData(
                      appOverviewCardsData.remoteClusterCount !== -1 ||
                      appOverviewCardsData.localClusterDeploy
                        ? 0
                        : -1,
                      clusterCount,
                      '30%'
                    )}
                  </div>
                </div>

                <div className="details-item" id="resources-status">
                  <div className="details-item-title right-item">
                    {msgs.get(
                      'dashboard.card.overview.cards.cluster.resource.status',
                      locale
                    )}
                  </div>
                  <div className="details-item-content">
                    {this.renderData(
                      appOverviewCardsData.nodeStatuses,
                      this.createStatusIcons(appOverviewCardsData.nodeStatuses),
                      '30%'
                    )}
                  </div>
                </div>

                <div className="details-item">
                  <a
                    className="details-item-link"
                    id="app-search-link"
                    href={getUrl + appOverviewCardsData.targetLink}
                    target="_blank"
                    rel="noreferrer"
                  >
                    <div>
                      {msgs.get(
                        'dashboard.card.overview.cards.search.resource',
                        locale
                      )}
                      <ArrowRightIcon className="details-item-link-icon" />
                    </div>
                  </a>
                </div>
              </div>
            </AccordionContent>
          </AccordionItem>
        </Accordion>

        <div className="overview-cards-subs-section">
          {showSubCards && !disableBtn
            ? this.createSubsCards(appOverviewCardsData.subsList, locale)
            : ''}
          <Button
            className="toggle-subs-btn"
            variant="secondary"
            isDisabled={disableBtn}
            data-test-subscription-details={!disableBtn}
            onClick={() => this.toggleSubsBtn(showSubCards)}
          >
            {this.renderData(
              appOverviewCardsData.subsList,
              (showSubCards
                ? msgs.get(
                  'dashboard.card.overview.cards.subs.btn.hide',
                  locale
                )
                : msgs.get(
                  'dashboard.card.overview.cards.subs.btn.show',
                  locale
                )) + ` (${appOverviewCardsData.subsList.length})`,
              '70%'
            )}
          </Button>
        </div>
      </div>
    )
  }

  renderData = (checkData, showData, width) => {
    return checkData !== -1 ? (
      showData
    ) : (
      <Skeleton width={width} className="loading-skeleton-text" />
    )
  };

  createStatusIcons = nodeStatuses => {
    return (
      <React.Fragment>
        <div
          className="status-icon-container green-status"
          id="green-resources"
        >
          <img
            className="status-icon"
            alt="node-status-success"
            src={`${config.contextPath}/graphics/nodeStatusSuccess.svg`}
          />
          <div className="status-count">{nodeStatuses.green}</div>
        </div>
        <div
          className="status-icon-container orange-status"
          id="orange-resources"
        >
          <img
            className="status-icon"
            alt="node-status-pending"
            src={`${config.contextPath}/graphics/nodeStatusPending.svg`}
          />
          <div className="status-count">{nodeStatuses.orange}</div>
        </div>
        <div
          className="status-icon-container yellow-status"
          id="yellow-resources"
        >
          <img
            className="status-icon"
            alt="node-status-warning"
            src={`${config.contextPath}/graphics/nodeStatusWarning.svg`}
          />
          <div className="status-count">{nodeStatuses.yellow}</div>
        </div>
        <div className="status-icon-container red-status" id="red-resources">
          <img
            className="status-icon"
            alt="node-status-failure"
            src={`${config.contextPath}/graphics/nodeStatusFailure.svg`}
          />
          <div className="status-count">{nodeStatuses.red}</div>
        </div>
      </React.Fragment>
    )
  };

  createSubsCards = (subsList, locale) => {
    return subsList.map(sub => {
      if (sub.name) {
        return (
          <React.Fragment key={sub.id}>
            <div className="sub-card-container">
              <div className="sub-card-column add-right-border">
                <img
                  className="subs-icon"
                  alt="subscription-card-sub-name"
                  src={`${config.contextPath}/graphics/subCardSubName.svg`}
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.subs.label',
                      locale
                    )}
                  </div>
                  <span>{sub.name}</span>
                </div>
              </div>

              <div className="sub-card-column add-right-border">
                <img
                  className="subs-icon"
                  alt="subscription-card-repo-folder"
                  src={`${config.contextPath}/graphics/subCardRepoFolder.svg`}
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.repoResource.label',
                      locale
                    )}
                  </div>
                  <ChannelLabels
                    channels={[
                      {
                        type: sub.resourceType,
                        pathname: sub.resourcePath,
                        gitBranch: sub.gitBranch,
                        gitPath: sub.gitPath,
                        package: sub.package,
                        packageFilterVersion: sub.packageFilterVersion
                      }
                    ]}
                    locale={locale}
                  />
                </div>
              </div>

              <div className="sub-card-column">
                <img
                  className="subs-icon"
                  alt="subscription-card-time-window"
                  src={`${config.contextPath}/graphics/subCardTimeWindow.svg`}
                />
                <div className="sub-card-content">
                  <div className="sub-card-title">
                    {msgs.get(
                      'dashboard.card.overview.cards.timeWindow.label',
                      locale
                    )}
                  </div>
                  {!sub.timeWindowType ? (
                    <div
                      className="set-time-window-link"
                      tabIndex="0"
                      role={'button'}
                      onClick={this.toggleEditorTab.bind(this)}
                      onKeyPress={this.toggleEditorTab.bind(this)}
                    >
                      {msgs.get(
                        'dashboard.card.overview.cards.timeWindow.set.label',
                        locale
                      )}
                    </div>
                  ) : (
                    <TimeWindowLabels
                      timeWindow={{
                        subName: sub.name,
                        type: sub.timeWindowType,
                        days: sub.timeWindowDays,
                        timezone: sub.timeWindowTimezone,
                        ranges: sub.timeWindowRanges
                      }}
                      locale={locale}
                    />
                  )}
                </div>
              </div>
            </div>
          </React.Fragment>
        )
      }
      return ''
    })
  };

  toggleEditorTab = () => {
    document.getElementById('editor').click()
  };

  toggleSubsBtn = showSubCards => {
    this.setState({ showSubCards: !showSubCards })
    this.forceUpdate()
  };
}

OverviewCards.propTypes = {}

export default withLocale(connect(mapStateToProps)(OverviewCards))
