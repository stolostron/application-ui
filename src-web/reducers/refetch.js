/*******************************************************************************
 * Copyright (c) 2020 Red Hat, Inc.
 *******************************************************************************/

import { REFETCH_INTERVAL_UPDATE } from '../actions'

export const refetch = (state, action) => {
  switch (action.type) {
  case REFETCH_INTERVAL_UPDATE:
    return action.data
  default:
    // get the poll interval from RefreshTimeSelect component
    return Object.assign({}, state, {})
  }
}
