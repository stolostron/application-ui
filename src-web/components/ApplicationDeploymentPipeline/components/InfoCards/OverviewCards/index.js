/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import R from 'ramda'
import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { SkeletonText } from 'carbon-components-react'
import resources from '../../../../../../lib/shared/resources'
import { RESOURCE_TYPES } from '../../../../../../lib/shared/constants'
import { fetchResources } from '../../../../../actions/common'
import { bindActionCreators } from 'redux'
import * as Actions from '../../../../../actions'
import msgs from '../../../../../../nls/platform.properties'
import {
  getNumIncidents,
  getSubscriptionDataOnHub,
  getSubscriptionDataOnManagedClustersSingle,
  getPodData,
  getIncidentsData,
  concatDataForTextKey,
  concatDataForSubTextKey
} from '../utils'
import { getSearchLinkForOneApplication } from '../../../../common/ResourceOverview/utils'
import config from '../../../../../../lib/shared/config'

/* eslint-disable react/prop-types */

resources(() => {
  require('../style.scss')
})

const mapDispatchToProps = dispatch => {
  return {
    fetchApplications: () =>
      dispatch(fetchResources(RESOURCE_TYPES.QUERY_APPLICATIONS)),
    actions: bindActionCreators(Actions, dispatch)
  }
}

const mapStateToProps = state => {
  const {
    QueryApplicationList,
    CEMIncidentList,
    secondaryHeader,
    AppOverview
  } = state
  const isSingleApplicationView =
    R.pathOr([], ['breadcrumbItems'])(secondaryHeader).length === 2
  const enableCEM = AppOverview.showCEMAction
  return {
    QueryApplicationList,
    CEMIncidentList,
    isSingleApplicationView,
    enableCEM
  }
}

const getOverviewCardsData = (
  QueryApplicationList,
  CEMIncidentList,
  isSingleApplicationView,
  enableCEM,
  applicationName,
  applicationNamespace,
  targetLinkForSubscriptions,
  targetLinkForClusters,
  targetLinkForPods,
  locale
) => {
  // All functions will return -1 if data or data.items is undefined... this will be used for skeleton text load bar
  const subscriptionDataOnHub = getSubscriptionDataOnHub(
    QueryApplicationList,
    isSingleApplicationView,
    applicationName,
    applicationNamespace
  )
  var subscriptionDataOnManagedClusters = getSubscriptionDataOnManagedClustersSingle(
    QueryApplicationList,
    applicationName,
    applicationNamespace
  )
  const podData = getPodData(
    QueryApplicationList,
    applicationName,
    applicationNamespace
  )
  const incidents = getNumIncidents(CEMIncidentList)
  const incidentsData = getIncidentsData(CEMIncidentList)

  const result = [
    {
      msgKey:
        subscriptionDataOnHub.total === 1
          ? msgs.get('dashboard.card.deployment.subscription', locale)
          : msgs.get('dashboard.card.deployment.subscriptions', locale),
      count: subscriptionDataOnHub.total,
      targetLink:
        subscriptionDataOnHub.total === 0 ? '' : targetLinkForSubscriptions,
      textKey: msgs.get('dashboard.card.deployment.subscriptions.text', locale),
      subtextKeyFirst: concatDataForSubTextKey(
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        subscriptionDataOnHub.total,
        subscriptionDataOnHub.noStatus,
        subscriptionDataOnHub.noStatus,
        msgs.get('dashboard.card.deployment.noStatus', locale)
      )
    },
    {
      msgKey:
        subscriptionDataOnManagedClusters.clusters === 1
          ? msgs.get('dashboard.card.deployment.managedCluster', locale)
          : msgs.get('dashboard.card.deployment.managedClusters', locale),
      count: subscriptionDataOnManagedClusters.clusters,
      targetLink:
        subscriptionDataOnManagedClusters.clusters === 0
          ? ''
          : targetLinkForClusters,
      textKey: concatDataForTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.total,
        msgs.get('dashboard.card.deployment.totalSubscription', locale),
        msgs.get('dashboard.card.deployment.totalSubscriptions', locale)
      ),
      subtextKeyFirst: concatDataForSubTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        subscriptionDataOnManagedClusters.clusters,
        subscriptionDataOnManagedClusters.noStatus,
        subscriptionDataOnManagedClusters.noStatus,
        msgs.get('dashboard.card.deployment.noStatus', locale)
      )
    },
    {
      msgKey:
        podData.total === 1
          ? msgs.get('dashboard.card.deployment.pod', locale)
          : msgs.get('dashboard.card.deployment.pods', locale),
      count: podData.total,
      targetLink: podData.total === 0 ? '' : targetLinkForPods,
      subtextKeyFirst: concatDataForSubTextKey(
        podData.total,
        podData.total,
        podData.running,
        msgs.get('dashboard.card.deployment.running', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        podData.total,
        podData.total,
        podData.failed,
        msgs.get('dashboard.card.deployment.failed.lowercase', locale)
      )
    }
  ]

  if (enableCEM) {
    result.push({
      msgKey:
        incidents === 1
          ? msgs.get('dashboard.card.deployment.incident', locale)
          : msgs.get('dashboard.card.deployment.incidents', locale),
      count: incidents,
      alert: incidents > 0 ? true : false,
      targetTab: incidents === 0 ? null : 2,
      subtextKeyFirst: concatDataForSubTextKey(
        incidents,
        incidents,
        incidentsData.priority1,
        msgs.get('dashboard.card.deployment.incidents.priority1', locale)
      ),
      subtextKeySecond: concatDataForSubTextKey(
        incidents,
        incidents,
        incidentsData.priority2,
        msgs.get('dashboard.card.deployment.incidents.priority2', locale)
      )
    })
  }
  return result
}

class OverviewCards extends React.Component {
  componentDidMount() {
    const { fetchApplications } = this.props

    fetchApplications()

    this.startPolling()
    document.addEventListener('visibilitychange', this.onVisibilityChange)
  }

  componentWillUnmount() {
    this.stopPolling()
    document.removeEventListener('visibilitychange', this.onVisibilityChange)
  }

  startPolling() {
    if (parseInt(config['featureFlags:liveUpdates'], 10) === 2) {
      var intervalId = setInterval(
        this.reload.bind(this),
        config['featureFlags:liveUpdatesPollInterval']
      )
      this.setState({ intervalId: intervalId })
    }
  }

  stopPolling() {
    if (this.state && this.state.intervalId) {
      clearInterval(this.state.intervalId)
    }
  }

  onVisibilityChange = () => {
    if (document.visibilityState === 'visible') {
      this.startPolling()
    } else {
      this.stopPolling()
    }
  };

  reload() {
    const { fetchApplications } = this.props
    fetchApplications()
  }

  render() {
    const {
      QueryApplicationList,
      CEMIncidentList,
      isSingleApplicationView,
      enableCEM,
      actions,
      selectedAppName,
      selectedAppNS
    } = this.props
    const { locale } = this.context
    const singleAppStyle = isSingleApplicationView ? ' single-app' : ''
    const applicationName = selectedAppName
    const applicationNamespace = selectedAppNS

    const targetLinkForSubscriptions = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'subscription'
    })
    const targetLinkForClusters = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'cluster'
    })
    const targetLinkForPods = getSearchLinkForOneApplication({
      name: encodeURIComponent(applicationName),
      showRelated: 'pod'
    })

    const overviewCardsData = getOverviewCardsData(
      QueryApplicationList,
      CEMIncidentList,
      isSingleApplicationView,
      enableCEM,
      applicationName,
      applicationNamespace,
      targetLinkForSubscriptions,
      targetLinkForClusters,
      targetLinkForPods,
      locale
    )

    return (
      <div className={'overview-cards-info' + singleAppStyle}>
        <InfoCards overviewCardsData={overviewCardsData} actions={actions} />
      </div>
    )
  }
}

const showCardData = (cardCount, width, className) => {
  return cardCount !== -1 ? (
    cardCount
  ) : (
    <SkeletonText width={width} className={className} />
  )
}

const InfoCards = ({ overviewCardsData, actions }) => {
  return (
    <React.Fragment>
      {Object.keys(overviewCardsData).map(key => {
        const card = overviewCardsData[key]
        const id = `${key}_overviewCardsData`
        var cemStatus = 'card-cem-disabled'
        if (overviewCardsData.length === 4) {
          cemStatus = 'card-cem-enabled'
        }
        const handleClick = (e, resource) => {
          if (resource.targetTab != null) {
            actions.setSelectedAppTab(resource.targetTab)
          } else if (resource.targetLink) {
            window.open(resource.targetLink, '_blank')
          }
        }
        const handleKeyPress = (e, resource) => {
          if (e.key === 'Enter') {
            handleClick(e, resource)
          }
        }

        return (
          <React.Fragment key={key}>
            <div
              id={id}
              key={card}
              className={
                card.targetLink || card.targetTab
                  ? `single-card clickable ${cemStatus}`
                  : `single-card ${cemStatus}`
              }
              role="button"
              tabIndex="0"
              onClick={e => handleClick(e, card)}
              onKeyPress={e => handleKeyPress(e, card)}
            >
              <div className={card.alert ? 'card-count alert' : 'card-count'}>
                {showCardData(
                  card.count,
                  '60%',
                  'loading-skeleton-text-header'
                )}
              </div>
              <div className="card-type">{card.msgKey}</div>
              <div className="card-text">
                {showCardData(card.textKey, '80%', 'loading-skeleton-text-lrg')}
              </div>
              {(card.subtextKeyFirst || card.subtextKeySecond) && (
                <div className="row-divider" />
              )}
              {card.subtextKeyFirst && (
                <div className="card-subtext">
                  {showCardData(
                    card.subtextKeyFirst,
                    '40%',
                    'loading-skeleton-text-sm'
                  )}
                </div>
              )}
              {card.subtextKeySecond && (
                <div className="card-subtext">
                  {showCardData(
                    card.subtextKeySecond,
                    '60%',
                    'loading-skeleton-text-med'
                  )}
                </div>
              )}
            </div>
            {key < Object.keys(overviewCardsData).length - 1 && (
              <div className="column-divider" />
            )}
          </React.Fragment>
        )
      })}
    </React.Fragment>
  )
}

InfoCards.propTypes = {
  actions: PropTypes.object,
  overviewCardsData: PropTypes.array
}

OverviewCards.propTypes = {}

export default connect(mapStateToProps, mapDispatchToProps)(OverviewCards)
