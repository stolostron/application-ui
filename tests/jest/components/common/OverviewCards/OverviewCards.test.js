/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *  Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
jest.mock("../../../../../lib/client/apollo-client", () => ({
  getSearchClient: jest.fn(() => {
    return null;
  }),
  get: jest.fn(resourceType => {
    //resourceType.list is always ApplicationsList
    return Promise.resolve({
      data: {
        applications: {
          "mortgage-app-default": {
            _uid: "local-cluster/5cd1d4c7-52aa-11ea-bf05-00000a102d26",
            name: "mortgage-app",
            namespace: "default",
            dashboard:
              "https://localhost:443/grafana/dashboard/db/mortgage-app-dashboard-via-federated-prometheus?namespace=default",
            clusterCount: 1,
            remoteSubscriptionStatusCount: {
              Subscribed: 1
            },
            podStatusCount: {
              Running: 1
            },
            hubSubscriptions: [
              {
                _uid: "local-cluster/5cdc0d8d-52aa-11ea-bf05-00000a102d26",
                status: "Propagated",
                channel: "default/mortgage-channel",
                __typename: "Subscription"
              }
            ],
            created: "2020-02-18T23:57:04Z",
            __typename: "Application"
          }
        }
      }
    });
  }),
  search: jest.fn(resourceType => Promise.resolve({ response: resourceType }))
}));
const React = require("../../../../../node_modules/react");

import OverviewCards from "../../../../../src-web/components/common/OverviewCards";

import { mount } from "enzyme";
import renderer from "react-test-renderer";
import * as reducers from "../../../../../src-web/reducers";

import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";
import { Provider } from "react-redux";

import configureMockStore from "redux-mock-store";
import { MockedProvider } from "react-apollo/test-utils";

import { reduxStoreAppPipelineWithCEM, serverProps } from "../../TestingData";

const preloadedState = window.__PRELOADED_STATE__;
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const middleware = [thunkMiddleware];

const store = createStore(
  combineReducers(reducers),
  preloadedState,
  composeEnhancers(applyMiddleware(...middleware))
);

const mockStore = configureMockStore(middleware);
const storeApp = mockStore(reduxStoreAppPipelineWithCEM);

window.open = () => {}; // provide an empty implementation for window.open

describe("OverviewCards", () => {
  it("OverviewCards makes apollo calls with success return", () => {
    renderer.create(
      <MockedProvider mocks={[]} addTypename={false}>
        <Provider store={storeApp}>
          <OverviewCards
            selectedAppName="mortgage-app"
            selectedAppNS="default"
            serverProps={serverProps}
          />
        </Provider>
      </MockedProvider>
    );
  });

  it("has functioning onclick, one app", () => {
    const wrapper = mount(
      <Provider store={storeApp}>
        <OverviewCards
          selectedAppName="mortgage-app"
          selectedAppNS="default"
          serverProps={serverProps}
        />
      </Provider>
    );

    wrapper.find({ id: "0_overviewCardsData" }).simulate("click");
    wrapper
      .find({ id: "0_overviewCardsData" })
      .simulate("keypress", { key: "Enter" });
  });

  it("OverviewCards renders correctly with data.", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <OverviewCards selectedAppName="app1" selectedAppNS="default" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("OverviewCards successfully handles change in props to refetch", () => {
    const wrapper = mount(
      <Provider store={storeApp}>
        <OverviewCards
          selectedAppName="mortgage-app"
          selectedAppNS="default"
          serverProps={serverProps}
        />
      </Provider>
    );

    // change interval to 30000
    wrapper.setState({ refetch: { doRefetch: false, interval: 30000 } });
    // manual refetch data
    wrapper.setState({ refetch: { doRefetch: true, interval: 30000 } });
  });
});
