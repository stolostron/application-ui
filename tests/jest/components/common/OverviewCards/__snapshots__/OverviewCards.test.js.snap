// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewCards OverviewCards renders correctly with data. 1`] = `
<div
  className="overview-cards-container"
>
  <dl
    aria-label=""
    className="pf-c-accordion"
  >
    <dt>
      <button
        aria-expanded={true}
        className="pf-c-accordion__toggle pf-m-expanded overview-cards-details-section"
        id="details-card"
        onClick={[Function]}
      >
        <span
          className="pf-c-accordion__toggle-text"
        >
          Details
        </span>
        <span
          className="pf-c-accordion__toggle-icon"
        >
          <svg
            aria-hidden={true}
            aria-labelledby={null}
            fill="currentColor"
            height="1em"
            role="img"
            style={
              Object {
                "verticalAlign": "-0.125em",
              }
            }
            viewBox="0 0 256 512"
            width="1em"
          >
            <path
              d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            />
          </svg>
        </span>
      </button>
    </dt>
    <dd
      aria-label=""
      className="pf-c-accordion__expanded-content pf-m-expanded"
      hidden={false}
      id=""
    >
      <div
        className="pf-c-accordion__expanded-content-body"
      >
        <div
          className="details-col  add-right-border"
          id="left-col"
        >
          <div
            className="details-item"
          >
            <div
              className="details-item-title left-item"
            >
              Name
            </div>
            <div
              className="details-item-content"
            >
              app1
            </div>
          </div>
          <div
            className="details-item"
          >
            <div
              className="details-item-title left-item"
            >
              Namespace
            </div>
            <div
              className="details-item-content"
            >
              default
            </div>
          </div>
          <div
            className="details-item"
          >
            <div
              className="details-item-title left-item"
            >
              Created
            </div>
            <div
              className="details-item-content"
            >
              <div
                className="pf-c-skeleton loading-skeleton-text"
                style={
                  Object {
                    "--pf-c-skeleton--Height": undefined,
                    "--pf-c-skeleton--Width": "30%",
                  }
                }
              >
                <span
                  className="pf-u-screen-reader"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="details-col"
          id="righ-col"
        >
          <div
            className="details-item"
          >
            <div
              className="details-item-title right-item"
            >
              Clusters
            </div>
            <div
              className="details-item-content"
            >
              <div
                className="pf-c-skeleton loading-skeleton-text"
                style={
                  Object {
                    "--pf-c-skeleton--Height": undefined,
                    "--pf-c-skeleton--Width": "30%",
                  }
                }
              >
                <span
                  className="pf-u-screen-reader"
                />
              </div>
            </div>
          </div>
          <div
            className="details-item"
            id="resources-status"
          >
            <div
              className="details-item-title right-item"
            >
              Cluster resource status
            </div>
            <div
              className="details-item-content"
            >
              <div
                className="pf-c-skeleton loading-skeleton-text"
                style={
                  Object {
                    "--pf-c-skeleton--Height": undefined,
                    "--pf-c-skeleton--Width": "30%",
                  }
                }
              >
                <span
                  className="pf-u-screen-reader"
                />
              </div>
            </div>
          </div>
          <div
            className="details-item"
          >
            <a
              className="details-item-link"
              href="/multicloud/search?filters={\\"textsearch\\":\\"kind%3Aapplication%20name%3Aapp1%20namespace%3Adefault\\"}"
              id="app-search-link"
              rel="noreferrer"
              target="_blank"
            >
              <div>
                Search resource
                <svg
                  aria-hidden={true}
                  aria-labelledby={null}
                  className="details-item-link-icon"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style={
                    Object {
                      "verticalAlign": "-0.125em",
                    }
                  }
                  viewBox="0 0 448 512"
                  width="1em"
                >
                  <path
                    d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
                    transform=""
                  />
                </svg>
              </div>
            </a>
          </div>
        </div>
      </div>
    </dd>
  </dl>
  <div
    className="overview-cards-subs-section"
  >
    
    <button
      aria-disabled={true}
      aria-label={null}
      className="pf-c-button pf-m-secondary pf-m-disabled toggle-subs-btn"
      data-ouia-component-id="OUIA-Generated-Button-secondary-3"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe={true}
      data-test-subscription-details={false}
      disabled={true}
      onClick={[Function]}
      role={null}
      tabIndex={null}
      type="button"
    >
      <div
        className="pf-c-skeleton loading-skeleton-text"
        style={
          Object {
            "--pf-c-skeleton--Height": undefined,
            "--pf-c-skeleton--Width": "70%",
          }
        }
      >
        <span
          className="pf-u-screen-reader"
        />
      </div>
    </button>
  </div>
</div>
`;
